### DISCUSSION TITLE
More complete examples for _creating_ JSON elements?

### DISCUSSION THREAD

**Discussion by madscientist** [nlohmann/json#2886] [General]
Is there somewhere to go to see more comprehensive documentation on how to use this library to **create** JSON?  For example writing `to_json()` methods for more complex situations?

In particular I'm looking for best practice advice and performance pitfalls to avoid when generating `to_json()`: for example how to avoid making multiple copies of data (copying one time is fine with me: I'm not trying to avoid _all_ copies, just the redundant ones).

The examples in the documentation show parsing a static string into a JSON object, which is neat, but I suspect most people don't have a string containing all the data to be converted.  Instead, they need to build up a complex structure of JSON from another complex structure of C++ content, in a way that avoids recopying things multiple times.

For example I have a to_json() method I'm writing for a data structure and it involves returning an array of objects, where each object contains a mix of simple values as well as more complex values (arrays and objects), etc.  At no point do I ever have a full structure (like a vector or a map) that I can just pass to the JSON constructor (at least not without copying all my data into a separate structure which I obviously don't want to do).  I need to build up these individual JSON elements and add them into the top-level element incrementally.

Is the best practice to create a stand-alone json object, then add it into a higher level json object via `std::move()`?  Something like:

```
void to_json(nlohmann::json& json, const MyData& data)
{
    for (int n = 0; n < data.next; ++n) {
        nlohmann::json jevent;
        jevent["num"] = n;

        jevent["type"] = data.gettype();

        nlohmann::json jvalues;
        for (int i = 0; i < data.value_count(); ++i) {
            jvalues.emplace_back(data.getvalue(i));
        }
        jevent["values"] = std::move(jvalues);

        json.emplace_back(std::move(jevent));
    }
}
```

Is this enough to prevent a lot of extra copying?  Or, are there other ways to do this that are better?

I think the docs spend a lot of time discussing various ways to parse JSON but don't have a lot of examples or things to say about ways to create JSON.  But maybe I'm missing it?

**Comment by nlohmann** [comment]
There is no explicit documentation for how code inside a `to_json` function should look like. The API of `basic_json` is designed to resemble that of a `std::vector` or `std::map`, so any best practices there should also apply to this library, including using move constructors. Any concrete proposal?

**Reply by gregmarr** [reply]
That example is pretty optimal as it's written.  I would probably replace `jevent["values"] = std::move(jvalues);` with `jevent.emplace("values", std::move(jvalues));`.  I tried to write a small update using `reserve` to prevent reallocation of the array as it grew, but I guess that isn't exposed.

**Reply by gregmarr** [reply]
Another option, that uses the implicit conversion to array on first `emplace_back`:

        nlohmann::json &jvalues = jevent["values"];
        for (int i = 0; i < data.value_count(); ++i) {
            jvalues.emplace_back(data.getvalue(i));
        }

**Reply by gregmarr** [reply]
Another pass, using implicit object this time:

    for (int n = 0; n < data.next; ++n) {
        nlohmann::json &jevent = json.emplace_back();
        jevent["num"] = n;
        jevent["type"] = data.gettype();

        nlohmann::json &jvalues = jevent["values"];
        for (int i = 0; i < data.value_count(); ++i) {
            jvalues.emplace_back(data.getvalue(i));
        }
    }

**Reply by madscientist** [reply]
It doesn't necessarily need to be in the context of creating to_json() although that's a likely place where people will want to generate JSON.  It can just be normal open-coding to translate a complex set of C++ structures into a JSON object, where we don't have all the data as either a static string or as vector or map STL values, that could just be passed directly to a json ctor.  And, where we need to call some functions to generate some parts, then hook them into a larger json object.

The examples by @gregmarr are great, just the kind of discussion I was looking for.  In general I don't think people will immediately think of creating json objects for the the lowest levels of the result and std::move()ing them up into higher levels: that's not the first way I thought to do it, although I did eventually.  And if they're not considering it carefully they will invoke a lot of copying by not using std::move().

And, the idea of getting a reference never occurred to me; that idiom is probably something that will make things easier in many cases where we can get a reference to a location in a higher-level json then pass that around to other methods to fill out.  Would it work as well for arrays (what would you add?  Just an empty json object?)


So, I was just hoping that the docs could grow a section on methods for efficient, step-wise construction of json objects that shows examples like and gives a heads-up about things that might seem like a reasonable idea at first, but actually result in some undesirable behavior.  That would give newcomers a nice leg-up.

**Reply by madscientist** [reply]
Oh I guess you showed it working with an array -- doh! :)

**Reply by gregmarr** [reply]
Pulling out the pieces from my messages:

This works with a null json and converts it into an array, or adds to an existing array, and gets a reference to the newly added null json:

    nlohmann::json &jevent = json.emplace_back();

This works with a null json and converts it into an object, or adds a new entry to an existing object, and gets a reference to the newly added null json:

    nlohmann::json &jvalues = jevent["values"];

Assumptions: `json` and `jevent` are not `const`, and `"values"` doesn't already exist in the `jevent` object.

**Reply by madscientist** [reply]
Thanks @gregmarr  If you can pull some of this content into a separate answer rather than a reply I will mark it accepted.  Or I can do it if you prefer.

However IMO it would be great for this to appear as a section in the docs rather than/ in addition to in the FAQ... there are so many FAQs here that it can be daunting to find things (my searches yielded 10-15 matches but none of them were really what I was looking for).

Cheers, and thanks for the software!

**Reply by gregmarr** [reply]
@madscientist It would probably be better for you to pull together what you find useful, as I've been using this library for years, and no longer really have a good vision for what new users need to get going with it.
